src/ssl/ServerBump.h

```diff
@@ -49,6 +49,9 @@ class ServerBump
     /// whether we are currently performing one of the given processing steps
     bool at(const BumpStep step1, const BumpStep step2) const { return at(step1) || at(step2); }
 
+    /// the bumping mode at given step
+    BumpMode bumpModeAtStep(const BumpStep stp);
```

Comment:
> It is clear that after "ssl_bump splice step2" matches, bumpModeAtStep(tlsBump2) will always return bumpSplice, during step2 and during step3 (if any)...



> 
> However, that covers just basic use cases. What does "bumping mode at stepN" really mean (in all use cases)?
> 
>     * Is it the action that matched explicit or implicit ssl_bump rule during that stepN?
> 
>     * And if that step has not evaluated (or has not finished evaluating) any ssl_bump rules, then the answer is bumpNone?

I think that "bumping mode at stepN" for a stepN which never evaluated does not make sense. But the answer bumpNone is not enough to describe it. bumpNone is a valid answer for step1 or step2 when these steps are evaluated.

> 
>     * And if stepN was _skipped_, then it is the action at the last non-skipped step (if it exists)?
> 
> 
> The last two suggestions sound odd together because only one of them looks at the previous step when things do not work out at the current step, but all these suggestions are based on the current PR code, AFAICT.
> 
> I am tempted to say that this method returns the last (explicit or implicit) ssl_bump rule action matched at or before the given step, but that is not how the method is implemented right now because of its special `step < stp` check. Perhaps that check should be removed? 

I agree this check here is problematic. Not sure if it must removed or replaced with something else. 

> If it is removed, the method should be renamed to something like lastBumpAction().

Well.  A such method will not work well. We are running in the danger to hit again the bug we are trying to fix here.

We are using this method in a case we are entered step3, but we need the decision at step2.
Imagine the case the FwdState calls a PeekingPeerConnector object, which runs acls at step3 sets the ServerBump::act.step3 (the action at bumping step3) and then exits.  The FwdState starts again a new PeekingPeerConnector object. A lastBumpAction method will return previous step3 action/decision.

OK, we can reset on failures the ServerBump::act.step3 value, eg if a PeekingPeerConnector fails, or before we start a new PeekingPeerConnector,  but I think this is will cause more problems.

What about to throw inside ServerBump::bumpModeAtStep if the step is not evaluated? Or define a bumpUnknown/bumpNotExist or other bumping mode?

Reply:
> > It is clear that after "ssl_bump splice step2" matches, bumpModeAtStep(tlsBump2) will always return bumpSplice, during step2 and during step3 (if any)...
> 
> > However, that covers just basic use cases. What does "bumping mode at stepN" really mean (in all use cases)?
> > 
> > * Is it the action that matched explicit or implicit ssl_bump rule during that stepN?
> > * And if that step has not evaluated (or has not finished evaluating) any ssl_bump rules, then the answer is bumpNone?
 
Before we continue, please note that my bullets were based on the current Ssl::ServerBump::bumpModeAtStep() method implementation in the PR. If the bullets correctly reflect PR code but do not make sense, then we need to change the code...

> I think that "bumping mode at stepN" for a stepN which never evaluated does not make sense. 

Such a position implies that there should be an assert or similar measure somewhere that would prevent this method from returning a value when stepN has not been evaluated or has not finished evaluating. In other words, we can (and must) declare the caller buggy if the caller is asking questions that do not make sense (instead of returning some answer that is likely to be misinterpreted).

> But the answer bumpNone is not enough to describe it. bumpNone is a valid answer for step1 or step2 when these steps are evaluated.

I agree!

I do not know yet whether I agree with your assessment that the method call itself does not make sense for not-started/unfinished stepN, but I know that if I agree with that, then the code has to change.


> > I am tempted to say that this method returns the last (explicit or implicit) ssl_bump rule action matched at or before the given step, but that is not how the method is implemented right now because of its special `step < stp` check. Perhaps that check should be removed?
 
> I agree this check here is problematic. Not sure if it must removed or replaced with something else.
 
> > If it is removed, the method should be renamed to something like lastBumpAction().
 
> Well. A such method will not work well. We are running in the danger to hit again the bug we are trying to fix here.

I share your concern, but I hope we can address it, in part, by changing step3 boundaries (already done) and, in part, by resetting act.stepN+ values when Ssl::ServerBump::step is set to N. If we decide to (re)try stepN, then act.stepN (and later) values should be reset because from the current transaction state point of view they have not happened yet. And there should be already code that sets the right Ssl::ServerBump::step value. We just need to make sure that code is called unconditionally, for both first attempts and retries.

Instead of resetting, we could _delay_ setting stepN until we know that stepN has fully succeeded. That does not work well AFAICT because the decision to retry stepN may come from a layer above the layer that completed TLS manipulations for stepN. In other words, stepN may be successful from SslBump point of view, but may have to be retried later anyway from HTTP/FwdState point of view. We cannot delay storing act.stepN value beyond the SslBump involvement, of course.
 
> OK, we can reset on failures the ServerBump::act.step3 value [...] before we start a new PeekingPeerConnector, but I think this is will cause more problems.
 
What kind of problems? These "act" values are never used as a complete history of what has happened, right? We can make sure that there are no accesses to those values that are not guarded by the "current step number" value. And the Ssl::ServerBump::step setter method will make sure that any stale values are reset.

> What about to throw inside ServerBump::bumpModeAtStep if the step is not evaluated?

A failed Must() or asset() would simply indicate a bug in our code. They are better than a wrong return value, of course, but they are still bad. I hope we can design the code so that we have no reason to believe that the caller can accidentally ask the wrong question (and then safely assert that).

> Or define a bumpUnknown/bumpNotExist or other bumping mode?

The problem is that the method caller (and all the other code that checks/iterates over any bumpX values!) would need to handle that special value as well. That is a lot more complex than resetting on retries AFAICT (because retries have to have some kind of "reset" code for other natural reasons; we will be just adding one more reset action to that code).

To summarize, I think we should:

1. Move Ssl::ServerBump::step into a private lastStep_ field.
2. Provide an Ssl::ServerBump::startStep(n) method that sets the lastStep_ field and clears the corresponding act.stepN.
3. Refactor Ssl::ServerBump::bumpModeAtStep(step) to look something like this:
```C++
assert(step <= lastStep_); // the step we are asked about has started or completed
switch(step) {
    case XactionStep::tlsBump3:
        if (act.step3 != bumpNone) return act.step3;
        // [fallthrough]]
    case XactionStep::tlsBump2:
        if (act.step2 != bumpNone) return act.step2;
        // [fallthrough]]
    case XactionStep::tlsBump1:
        assert(act.step1 != bumpNone || step == XactionStep::tlsBump1);
        return act.step1;
    default:
        assert(!"this method is meant for SslBump steps only");
        return bumpEnd;
}
assert(!"unreachable code");
return bumpEnd;
```
4. Rename bumpModeAtStep() to lastBumpAction() and, if possible, remove its `step` parameter, using the `lastStep_` instead.
5. Consider using bumpEnd as act.stepN initializers, adjusting the bumpModeAtStep() conditions accordingly.

Reply:
commit adbbf252
Note the bumping step3 reset in FwdState::connectedToPeer method.